File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
 
        slow, fast = head, head

        # find middle
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # Reverse 2nd half
        curr = slow.next
        prev = None
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        slow.next = None # split 1st half
        first, second = head, prev        

        while first and second:
            nxt1 = first.next # 1,2
            nxt2 = second.next # 4,3

            # 1,2,3,4
            # 1,4,2,3
            first.next = second
            second.next = nxt1

            first = nxt1
            second = nxt2

'''
    1,2,3,4,5,6
    1,6,2,5,3,4

    first half
    reverse half
    
'''            

File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        // find mid
        ListNode *slow=head, *fast=head;
        while(fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // reverse 2nd half
        ListNode *prev=nullptr, *next;
        ListNode *curr=slow;
        while(curr){
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }        

        // mix
        ListNode* first = head;
        ListNode* second = prev;
        while(second->next) {
            ListNode* tmp = first->next;
            first->next = second;
            first = tmp;
            tmp = second->next;
            second->next = first;
            second = tmp;
        }
    }
};